### PR TITLE
Speed Up Cellular Automaton implementation

### PR DESCRIPTION
**Pull request by VinInn**
Most purely technical changes that produces no regression and speedup CA by about a factor 2
We took the opportunity to fix a small imperfection in the use of the radius the now is w/r/t the beam spot.  The effect is mostly to sharpen the pt cut. This change produces minor regressions.

Timing for PU200 ttbar  phase2
```
900
grep TimeRe seedCA.log | cut -c 35-120 | sort -n -r | grep StepHit | head -n2
     0.363026  initialStepHitQuadruplets
     0.067059  initialStepHitDoublets
this PR
grep TimeRe seedCAVN7.log | cut -c 35-120 | sort -n -r | grep StepHit | head -n2
     0.199053  initialStepHitQuadruplets
     0.058190  initialStepHitDoublets
```
Timing at PU50 PhaseI
```
900
grep TimeRe caOri.log | cut -c 35-120 | sort -n -r | grep QuadStepHit | head -n4
     0.139890  lowPtQuadStepHitQuadruplets
     0.072600  detachedQuadStepHitQuadruplets
     0.012884  lowPtQuadStepHitDoublets
     0.009104  detachedQuadStepHitDoublets

This PR
grep TimeRe caVecet3.log  | cut -c 35-120 | sort -n -r | grep QuadStepHit | head -n4
     0.094063  lowPtQuadStepHitQuadruplets
     0.045189  detachedQuadStepHitQuadruplets
     0.012195  lowPtQuadStepHitDoublets
     0.008520  detachedQuadStepHitDoublets
```

MTV for ttbar PU0
http://innocent.home.cern.ch/innocent/RelVal/vectCA3/plots_highPurity/effandfake1.pdf
where blue is 900, red is technical only and black is this PR

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by VinInn** [comment]
I hardly believe that this PR is the cause of the test failure
Would have been much more useful to have a list of the failed AddOn

**Comment by davidlange6** [comment]
yea, but if you solve everything noted in the comment, you'll likely be ok - two 2017/2018 workflows failed.. 10042 and 10824

Module: CAHitQuadrupletEDProducer (crashed)

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18002/18576/runTheMatrix-results/10042.0_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step2_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

certainly related..

**Comment by VinInn** [comment]
Interesting, as I have run on three different datasets and never crashed

**Comment by makortel** [comment]
The segfault is in HLT, could there be some special case that is not treated properly now?

**Comment by VinInn** [comment]
maybe HLT takes a different path? (there is code that is copy/paste four times...)
Will known soon

**Comment by VinInn** [comment]
Not obvious what is doing in HLT, we need to check in details.
It does not seem to do what pretended by design (building a single CA for all regions).
For next PR: for the time being just reproduce previous behavior...

**Comment by VinInn** [comment]
@cmsbuild, please test

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h [hunk 1]
```diff
@@ -10,286 +10,291 @@
...
+  using Hit = RecHitsSortedInPhi::Hit;
+  using CAntuple = std::vector<unsigned int>;
+  using CAntuplet = std::vector<unsigned int>;
+  using CAColl = std::vector<CACell>;
+  using CAStatusColl = std::vector<CACellStatus>;
+  
+  
+  CACell(const HitDoublets* doublets, int doubletId, const int innerHitId, const int outerHitId) :
+    theDoublets(doublets), theDoubletId(doubletId)
+    ,theInnerR(doublets->rv(doubletId, HitDoublets::inner))
```
Comment (VinInn) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: this is different that original (now w/r/t/ beamspot)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/18002	80b67be
- code change is in line with the description; affecting 2017/2018 setups now where CA is enabled
- jenkins tests pass and show small differences in 2017/2018 workflows
- local test with more events confirm little change in MTV plots and essentially no effect on downstream quantities
- speedup is evident, the target reference workflow 2017 ttbar PU35 (10224 ) shows overall less than 1% speedup with significant decrease in CA modules (consistent with what was posted in the PR description)
```
step2:
2.18 ms/ev ->         0.80 ms/ev hltIter2PFlowPixelHitTriplets
36.89 ms/ev ->        21.37 ms/ev hltPixelTracksHitQuadruplets
10.01 ms/ev ->         5.91 ms/ev hltIter1PFlowPixelHitQuadruplets

step3:
75.90 ms/ev ->        48.32 ms/ev detachedQuadStepHitQuadruplets
51.65 ms/ev ->        33.70 ms/ev initialStepHitQuadruplets
159.75 ms/ev ->       107.67 ms/ev lowPtQuadStepHitQuadruplets
49.14 ms/ev ->        34.04 ms/ev detachedTripletStepHitTriplets
51.97 ms/ev ->        36.43 ms/ev initialStepHitQuadrupletsPreSplitting
103.96 ms/ev ->        77.96 ms/ev lowPtTripletStepHitTriplets
32.08 ms/ev ->        24.19 ms/ev highPtTripletStepHitTriplets
Total in reco: total: 15817 -> 15762	-0.4%
   *Step* modules: 7416 -> 7308        -1.5%
```

**Comment by davidlange6** [comment]
+1

**Pull request merged**